File: sapp/sapp/backend-sapp/src/services/starpayService.ts
/**
 * StarPay Service
 * Handles integration with StarPay API for virtual card purchases
 */

// Import types from dedicated type files
import type {
  CardOrderRequest,
  PaymentInfo,
  PricingInfo,
  CardOrderResponse,
  CardDetails,
  OrderStatusResponse,
  PriceResponse,
  OrderStatus,
  StarPayError,
} from '../types/starpay.types.js';

// Import constants
import { STARPAY_API_URL, STARPAY_LIMITS } from '../constants/api.js';

// Import utilities
import { isValidEmail } from '../utils/validation.js';

// Re-export types for consumers
export type {
  CardOrderRequest,
  PaymentInfo,
  PricingInfo,
  CardOrderResponse,
  CardDetails,
  OrderStatusResponse,
  PriceResponse,
  OrderStatus,
  StarPayError,
};

class StarPayService {
  private readonly baseUrl: string;
  private readonly apiKey: string;

  constructor() {
    this.baseUrl = process.env.STARPAY_API_URL || STARPAY_API_URL;
    this.apiKey = process.env.STARPAY_API_KEY || '';

    if (!this.apiKey) {
      console.warn('[StarPayService] Warning: STARPAY_API_KEY not configured');
    }
  }

  /**
   * Create a new card order
   */
  async createOrder(request: CardOrderRequest): Promise<CardOrderResponse> {
    this.validateOrderRequest(request);

    const response = await this.makeRequest<CardOrderResponse>(
      '/api/v1/cards/order',
      {
        method: 'POST',
        body: JSON.stringify({
          amount: request.amount,
          cardType: request.cardType,
          email: request.email,
        }),
      }
    );

    return response;
  }

  /**
   * Check order status
   */
  async getOrderStatus(orderId: string): Promise<OrderStatusResponse> {
    if (!orderId || orderId.trim().length === 0) {
      throw new Error('INVALID_ORDER_ID');
    }

    const response = await this.makeRequest<OrderStatusResponse>(
      `/api/v1/cards/order/status?orderId=${encodeURIComponent(orderId)}`,
      { method: 'GET' }
    );

    return response;
  }

  /**
   * Get pricing breakdown for an amount (without creating order)
   */
  async getPrice(amount: number): Promise<PriceResponse> {
    if (amount < STARPAY_LIMITS.MIN_AMOUNT || amount > STARPAY_LIMITS.MAX_AMOUNT) {
      throw new Error('INVALID_AMOUNT');
    }

    const response = await this.makeRequest<PriceResponse>(
      `/api/v1/cards/price?amount=${amount}`,
      { method: 'GET' }
    );

    return response;
  }

  /**
   * Validate card order request
   */
  private validateOrderRequest(request: CardOrderRequest): void {
    // Validate amount
    if (
      typeof request.amount !== 'number' ||
      request.amount < STARPAY_LIMITS.MIN_AMOUNT ||
      request.amount > STARPAY_LIMITS.MAX_AMOUNT
    ) {
      throw new Error('INVALID_AMOUNT');
    }

    // Validate card type
    if (!['visa', 'mastercard'].includes(request.cardType)) {
      throw new Error('INVALID_CARD_TYPE');
    }

    // Validate email using shared utility
    if (!request.email || !isValidEmail(request.email)) {
      throw new Error('INVALID_EMAIL');
    }
  }

  /**
   * Make authenticated request to StarPay API
   */
  private async makeRequest<T>(endpoint: string, options: RequestInit): Promise<T> {
    if (!this.apiKey) {
      throw new Error('STARPAY_NOT_CONFIGURED');
    }

    const url = `${this.baseUrl}${endpoint}`;

    const response = await fetch(url, {
      ...options,
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
        'Content-Type': 'application/json',
        ...options.headers,
      },
    });

    const data = (await response.json()) as T & { error?: string; message?: string };

    if (!response.ok) {
      const errorCode = data.error || 'UNKNOWN_ERROR';
      const errorMessage = data.message || 'An unknown error occurred';

      console.error(`[StarPayService] API error: ${errorCode} - ${errorMessage}`);

      // Map StarPay error codes to our error codes
      switch (response.status) {
        case 401:
          throw new Error('UNAUTHORIZED');
        case 403:
          throw new Error('ACCOUNT_SUSPENDED');
        case 404:
          throw new Error('ORDER_NOT_FOUND');
        case 400:
          throw new Error(errorCode);
        default:
          throw new Error('INTERNAL_ERROR');
      }
    }

    return data as T;
  }

  /**
   * Check if service is configured
   */
  isConfigured(): boolean {
    return !!this.apiKey;
  }
}

// Export singleton instance
export const starpayService = new StarPayService();

File: sapp/Sapp-ios/Core/StarPay/StarPayService.swift
import Foundation
import Combine

// MARK: - StarPay Service

@MainActor
final class StarPayService: ObservableObject {

    // MARK: - Singleton

    static let shared = StarPayService()

    // MARK: - Configuration

    private let baseURL: String
    private let session: URLSession

    // MARK: - Published State

    @Published private(set) var isServiceAvailable: Bool = false
    @Published private(set) var currentOrder: CardOrderResponse?
    @Published private(set) var orderStatus: OrderStatusResponse?

    // MARK: - Initialization

    private init() {
        self.baseURL = AppConfig.sappAPIURL

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 60
        self.session = URLSession(configuration: config)

        print("[StarPayService] Using API URL: \(baseURL)")
    }

    // MARK: - Service Status

    /// Check if card service is available
    func checkServiceStatus() async throws -> Bool {
        let endpoint = "/api/sapp/starpay/status"
        guard let url = URL(string: baseURL + endpoint) else {
            throw StarPayError.networkError(URLError(.badURL))
        }

        do {
            let (data, _) = try await session.data(from: url)
            let response = try JSONDecoder().decode(StarPayServiceStatusResponse.self, from: data)
            isServiceAvailable = response.available
            return response.available
        } catch {
            isServiceAvailable = false
            throw StarPayError.networkError(error)
        }
    }

    // MARK: - Create Order

    /// Create a new card order
    func createOrder(amount: Double, cardType: CardType, email: String) async throws -> CardOrderResponse {
        let endpoint = "/api/sapp/starpay/order"
        guard let url = URL(string: baseURL + endpoint) else {
            throw StarPayError.networkError(URLError(.badURL))
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "amount": amount,
            "cardType": cardType.rawValue,
            "email": email
        ]

        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        do {
            let (data, response) = try await session.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse else {
                throw StarPayError.unknown("Invalid response")
            }

            let decoded = try JSONDecoder().decode(StarPayOrderAPIResponse.self, from: data)

            if httpResponse.statusCode == 503 {
                throw StarPayError.serviceUnavailable
            }

            if httpResponse.statusCode == 400 {
                switch decoded.error {
                case "INVALID_AMOUNT":
                    throw StarPayError.invalidAmount
                case "INVALID_EMAIL":
                    throw StarPayError.invalidEmail
                case "INVALID_CARD_TYPE":
                    throw StarPayError.invalidCardType
                default:
                    throw StarPayError.unknown(decoded.message ?? "Invalid request")
                }
            }

            if httpResponse.statusCode == 401 {
                throw StarPayError.unauthorized
            }

            if httpResponse.statusCode == 403 {
                throw StarPayError.accountSuspended
            }

            guard decoded.success, let order = decoded.order else {
                throw StarPayError.unknown(decoded.message ?? "Failed to create order")
            }

            currentOrder = order
            return order
        } catch let error as StarPayError {
            throw error
        } catch let error as DecodingError {
            throw StarPayError.decodingError(error)
        } catch {
            throw StarPayError.networkError(error)
        }
    }

    // MARK: - Get Order Status

    /// Check the status of an existing order
    func getOrderStatus(orderId: String) async throws -> OrderStatusResponse {
        let endpoint = "/api/sapp/starpay/order/status"
        guard var urlComponents = URLComponents(string: baseURL + endpoint) else {
            throw StarPayError.networkError(URLError(.badURL))
        }

        urlComponents.queryItems = [URLQueryItem(name: "orderId", value: orderId)]

        guard let url = urlComponents.url else {
            throw StarPayError.networkError(URLError(.badURL))
        }

        do {
            let (data, response) = try await session.data(from: url)

            guard let httpResponse = response as? HTTPURLResponse else {
                throw StarPayError.unknown("Invalid response")
            }

            let decoded = try JSONDecoder().decode(StarPayStatusAPIResponse.self, from: data)

            if httpResponse.statusCode == 404 {
                throw StarPayError.orderNotFound
            }

            if httpResponse.statusCode == 401 {
                throw StarPayError.unauthorized
            }

            if httpResponse.statusCode == 503 {
                throw StarPayError.serviceUnavailable
            }

            guard decoded.success,
                  let orderId = decoded.orderId,
                  let status = decoded.status else {
                throw StarPayError.unknown(decoded.message ?? "Failed to get order status")
            }

            let statusResponse = OrderStatusResponse(
                orderId: orderId,
                status: status,
                payment: decoded.payment,
                pricing: decoded.pricing,
                card: decoded.card,
                expiresAt: decoded.expiresAt,
                completedAt: decoded.completedAt,
                failureReason: decoded.failureReason
            )

            orderStatus = statusResponse
            return statusResponse
        } catch let error as StarPayError {
            throw error
        } catch let error as DecodingError {
            throw StarPayError.decodingError(error)
        } catch {
            throw StarPayError.networkError(error)
        }
    }

    // MARK: - Get Pricing

    /// Get pricing breakdown for an amount (without creating order)
    func getPrice(amount: Double) async throws -> PriceResponse {
        let endpoint = "/api/sapp/starpay/price"
        guard var urlComponents = URLComponents(string: baseURL + endpoint) else {
            throw StarPayError.networkError(URLError(.badURL))
        }

        urlComponents.queryItems = [URLQueryItem(name: "amount", value: String(amount))]

        guard let url = urlComponents.url else {
            throw StarPayError.networkError(URLError(.badURL))
        }

        do {
            let (data, response) = try await session.data(from: url)

            guard let httpResponse = response as? HTTPURLResponse else {
                throw StarPayError.unknown("Invalid response")
            }

            let decoded = try JSONDecoder().decode(StarPayPriceAPIResponse.self, from: data)

            if httpResponse.statusCode == 400 {
                throw StarPayError.invalidAmount
            }

            if httpResponse.statusCode == 503 {
                throw StarPayError.serviceUnavailable
            }

            guard decoded.success,
                  let pricing = decoded.pricing,
                  let solPrice = decoded.solPrice,
                  let amountSol = decoded.amountSol else {
                throw StarPayError.unknown(decoded.message ?? "Failed to get pricing")
            }

            return PriceResponse(pricing: pricing, solPrice: solPrice, amountSol: amountSol)
        } catch let error as StarPayError {
            throw error
        } catch let error as DecodingError {
            throw StarPayError.decodingError(error)
        } catch {
            throw StarPayError.networkError(error)
        }
    }

    // MARK: - Clear State

    func clearState() {
        currentOrder = nil
        orderStatus = nil
    }
}
